### ISSUE TITLE
SqlSerDsc Issue : PowerShell DSC resource MSFT_SqlSetup failed to execute Set-TargetResource functionality with error message

### ISSUE THREAD

**Issue by baba-dude** [dsccommunity/SqlServerDsc#1273]
I am Trying to Install SqlServer on windows from ansible. Unfortunately i am stuck at this error.

PowerShell DSC resource MSFT_SqlSetup failed to execute Set-TargetResource functionality with error message:System.Exception: Test-TargetResource returned false after calling Set-TargetResource.

Can you Please suggest me why i am getting these kind of error.

My Playbook paraeters are as below.
```
- name: Install SQL Server
  win_dsc:
    resource_name: SQLSetup
    Action: Install
    UpdateEnabled: True
    SourcePath: C:\SQLInstall\SQLServer2017-DEV-x64-ENU
    InstanceName: "{{ mssql_instance_name }}"
    InstallSharedDir: "{{ mssql_installshared_path }}"
    InstallSharedwowDir: "{{ mssql_installsharedwow_path }}"
    InstanceDir: "{{ mssql_instance_path }}"
    InstallSQLDataDir: "{{ mssql_sqlinstalldata_path }}"
    SQLUserDBDir: "{{ mssql_sqluserdata_path }}"
    SQLUserDBLogDir: "{{ mssql_sqluserlog_path }}"
    SQLTempDBDir: "{{ mssql_sqltempDB_path }}"
    SQLTempDBLogDir: "{{ mssql_sqltempDBlog_path }}"
    Features: "{{ mssql_features }}"
    SQLCollation: "{{ mssql_collation }}"
    BrowserSvcStartupType: "{{ mssql_browsersvc_mode }}"
    SuppressReboot: "{{ mssql_suppress_reboot }}"
    # SQL Service Account
    SQLSvcAccount_username: NT Service\SQLSERVERAGENT
    SQLSvcAccount_password: Password@123
    # SQL Agent Service Account
    AgtSvcAccount_username: NT Service\SQLSERVERAGENT
    AgtSvcAccount_password: Password@123
    # SQL Analysis Services Account
    ASSvcAccount_username: NT Service\SQLSERVERAGENT
    ASSvcAccount_password: Password@123
    # Used when installing on a network path, comment out
    # SourceCredential_username: "{{ ansible_user }}"
    # SourceCredential_password: "{{ ansible_password }}"

    # System Admins
    SQLSysAdminAccounts: win1\baba
    # Analysis Services Admins (if installed)
    ASSysAdminAccounts: win1\baba
  tags: install_sql
```

Please advise..:

**Comment by johlju** [comment]
As the last step the `Set-TargetResource` verifies the installation by checking if the instance is present. The error message means that the `Test-TargetResource` could not find the instance installed.

Look at the SQL Server install log (the SUMMARY.txt) file to see if the SQL Server installation failed, and with what error message. 
Or run the configuration locally with verbose messages `Start-DscConfiguration -Path <config> -Wait -Verbose`. The verbose output might give a hint too.

**Comment by baba-dude** [comment]
Thanks for the reply. I am running this from ansible i.e i am automating the installation process. I am not sure where can i find the summary.txt file. Is there any other way we can debug it ?

**Comment by johlju** [comment]
You can read here about the [SQL server logs](https://docs.microsoft.com/en-us/sql/database-engine/install-windows/view-and-read-sql-server-setup-log-files?view=sql-server-2017)

I’m not familiar with Ansible so not sure that you can enable verbose logging there for the DSC resource. My suggestion was to run this one time manually with PowerShell to determine the cause so you after that can automate it correctly.

**Comment by johlju** [comment]
If you look at the Summary.txt file in the `%programfiles%\Microsoft SQL Server\nnn\Setup Bootstrap\Log` does it say that the installation failed?

**Comment by hanssonsebastian** [comment]
Having the same issue, also running with Ansible. All looks good after the install and the log says it Passed. Atm I'm just ignoring this error during setup.

The Summary.log says:
Overall summary:
  Final result:                  Passed
  Exit code (Decimal):           0
  Start time:                    2019-01-31 09:22:17
  End time:                      2019-01-31 09:29:14
  Requested action:              Install

**Comment by MarekSarad** [comment]
I am facing the same issue but I don't think the problem is with DSC SQLServer module but with implementation in Ansible. I have successfully installed SQL using DSC locally on the target server. When the same definition is applied using Ansible i get this error. But contrary to the @hanssonsebastian in my case the SQL is not installed at all. And I don't get anything in the SQL Installation logs.

**Comment by MarekSarad** [comment]
OK so I have followed this method (first answer https://stackoverflow.com/questions/54108728/powershell-passing-array-to-other-script-turns-to-system-hastable-fix) to try to troubleshoot. 

I have created this script: 

```
$params = @{
            InstanceName          = "INST2014"
            Features              = "SQLENGINE,SSMS,ADV_SSMS"
            SQLCollation          = "Latin1_General_CI_AS_KS_WS"
            #SQLSysAdminAccounts   = "TestAccount"
            InstallSharedDir      = "C:\Program Files\Microsoft SQL Server"
            InstallSharedWOWDir   = "C:\Program Files (x86)\Microsoft SQL Server"
            InstanceDir           = "C:\Program Files\Microsoft SQL Server"
            InstallSQLDataDir     = "D:\MSSQL\Data"
            SQLUserDBDir          = "D:\MSSQL\Data"
            SQLUserDBLogDir       = "D:\MSSQL\DataLogs"
            SQLTempDBDir          = "D:\MSSQL\TempDBData"
            SQLTempDBLogDir       = "D:\MSSQL\TempDBDataLogs"
            SQLBackupDir          = "D:\MSSQL\Backup"
            SourcePath            = "D:\Temp\2014 Enterprise x64"
            UpdateEnabled         = "False"            
            BrowserSvcStartupType = "Automatic"
            #PsDscRunAsCredential  = $cred
}

$params.GetType()

Invoke-DscResource -Name SQLSetup -ModuleName SQLServerDSC  -Verbose -Method Test -Property $params
```

When I execute it it works ok.

But as soon as I uncomment the #SQLSysAdminAccounts I get an error:

> VERBOSE: Perform operation 'Invoke CimMethod' with following parameters, ''methodName' = ResourceTest,'className' = MSFT_DSCLocalConfiguration
> Manager,'namespaceName' = root/Microsoft/Windows/DesiredStateConfiguration'.
> VERBOSE: An LCM method call arrived from computer DFS-SPTTST-720 with user sid S-1-5-21-3217640906-2541433269-1927226733-7701.
> **Convert property 'SQLSysAdminAccounts' value from type 'STRING' to type 'STRING[]' failed**
>  At line:23, char:2
>  Buffer:
> \\MSSQL\\TempDBData";
> };^
> insta
>     + CategoryInfo          : SyntaxError: (root/Microsoft/...gurationManager:String) [], CimException
>     + FullyQualifiedErrorId : MiClientApiError_Failed
>     + PSComputerName        : localhost
>  
> VERBOSE: Operation 'Invoke CimMethod' complete.
> VERBOSE: Time taken for configuration job to complete is 0.236 seconds

**Comment by MarekSarad** [comment]
looking further into DSC code it looks like PsDSCRunAsCredential is used to be added as default admin so I included that in the parameters. When I add the PSDSCRunAsCredential to the set of $params using this code above previous example:

```
if (-not $credential) {
    $password = "SomePass" | ConvertTo-SecureString -asPlainText -Force
$username = "Domain\InstallAccount"
[PSCredential] $credential = New-Object System.Management.Automation.PSCredential($username,$password)
}
```
Then I get another error:
```

Invoke-DscResource : Failed to serialize properties into CimInstance.
At C:\temp\TestSQLDSCvia_Invoke.ps1:36 char:1
+ Invoke-DscResource -Name SQLSetup -ModuleName SQLServerDSC  -Verbose  ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Invoke-DscResource], SerializationException
    + FullyQualifiedErrorId : System.Runtime.Serialization.SerializationException,Microsoft.PowerShell.DesiredStateConfiguration.Commands.In 
   vokeDscResourceMethodCommand
```
If I execute it without specifying Parameters initially only after prompt I get this: 

```
PS C:\Windows\system32> Invoke-DscResource -Name SQLSetup -ModuleName SQLServerDSC  -Verbose -Method Test
cmdlet Invoke-DscResource at command pipeline position 1
Supply values for the following parameters:
Property: $params
Invoke-DscResource : Cannot bind parameter 'Property'. Cannot convert the "$params" value of type "System.String" to type 
"System.Collections.Hashtable".
At line:1 char:1
+ Invoke-DscResource -Name SQLSetup -ModuleName SQLServerDSC  -Verbose  ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidArgument: (:) [Invoke-DscResource], ParameterBindingException
    + FullyQualifiedErrorId : CannotConvertArgumentNoMessage,Microsoft.PowerShell.DesiredStateConfiguration.Commands.InvokeDscResourceMethod 
   Command
```

**Comment by mdaniou** [comment]
@MarekSarad It is hard to understand your problem. The parameter SQLSysAdminAccounts expects an array, hence did you try this instead ?
`SQLSysAdminAccounts   = [Array]"TestAccount"`
Although I don't know how you can tell that your problem is the same as @baba-dude 's.
You might want to create an new issue.

**Comment by MarekSarad** [comment]
@mdaniou Thanks for you comment. please see my first response I think there is no problem with DSC as using only DSC on local machine I was able to install . I am in same situation I try to install SQL via Ansible using DSC and I am hitting same error. 

My other subsequent comments are from trying to troubleshoot that issue (and prove that DSC module is ok and problem probably is with how Ansible applies it). Using method mentioned here: https://stackoverflow.com/questions/54108728/powershell-passing-array-to-other-script-turns-to-system-hastable-fix 

You are correct that adding [Array]"TestAccount" solved my first problem and I can complete the Test without issue so pretty sure that problem is with the win_dsc module of Ansible not with the DSC. 

Thanks for help off to Ansible forum ;)

**Comment by MarekSarad** [comment]
I have open a new ticket on the Ansible Core GIT: https://github.com/ansible/ansible/issues/51543 
Leaving that here if anyone will stumble on this thread in the future.

**Comment by johlju** [comment]
Awesome that you guys solved this one! Seems when running this locally in DSC the engine is converting the string to string array automatically, but Ansible does not and passing it as a single string. 🤔 

> `SQLSysAdminAccounts = [Array]"TestAccount"`

Side note. A quicker way of writing this (in PowerShell) is `SQLSysAdminAccounts = @("TestAccount")`. But the result is the same.

**Comment by johlju** [comment]
@baba-dude Does this solve your issue as well?

**Comment by mdaniou** [comment]
@johlju 

> A quicker way of writing this (in PowerShell) is SQLSysAdminAccounts = @("TestAccount")

Is it ? ;)

**Comment by jborean93** [comment]
I think this can be closed as the issue seemed to be some incompatibilities with options being passed through. Now that https://github.com/ansible/ansible/pull/51927 has been merged, warnings generated by the DSC engine are now returned back on the Ansible output. In this particular case we would have seen the following error

```
TASK [install SQL Server] ******************************************************************************
task path: /Users/jborean/dev/module-tester/adhoc.yml:34
Using module file /Users/jborean/dev/ansible/lib/ansible/modules/windows/win_dsc.ps1
<127.0.0.1> ESTABLISH WINRM CONNECTION FOR USER: vagrant on PORT 55986 TO 127.0.0.1
EXEC (via pipeline wrapper)
 [WARNING]: [WIN-RUEMJHAT927]:                            [[SqlSetup]DirectResourceAccess] Setup exited
with code '-2068578302'. Please see the 'Summary.txt' log file in the 'Setup Bootstrap\Log' folder.

fatal: [vagrant-win]: FAILED! => {
    "changed": false,
    "module_version": "12.2.0.0",
    "msg": "PowerShell DSC resource MSFT_SqlSetup  failed to execute Set-TargetResource functionality with error message: System.Exception: Test-TargetResource returned false after calling Set-TargetResource. ",
    "reboot_required": false
}
```

This tells up the SQL setup.exe call failed with a non 0 exit code and to have a look at the Summary.txt file for more info on what went wrong. It might be nice to improve the error message for this scenario though, we can see at https://github.com/PowerShell/SqlServerDsc/blob/dev/DSCResources/MSFT_SqlSetup/MSFT_SqlSetup.psm1#L1586-L1650 this is the general workflow;

* Invoke setup.exe with the arguments specified
    * If rc is 3010 and we surpressed a reboot, write a warning saying a reboot is required
    * else if the rc is not 0, write a warning
    * else everything is good
* Check if a reboot is required and set some flags if that is the case
* If the setup was 0 or 3010, import the SQLPS module
* Finally test the target resource <-- this is where the error happens

The trouble is that we have a non-zero RC from the setup process and yet, apart from displaying a warning, the code continued on like everything was fine. I'm not sure if this was done on purpose but I would have assumed anything that isn't 0 or 3010 is an error and to fail saying so.

For anyone curious around the array/list behaviour for the Ansible win_dsc module. The code is able to detect if the DSC resource expects and array and automatically converts a string to the array or preserves the yaml/python list defined in the task. This was not the issue at hand here though.
